Feature: Allow using mdsthin for the MATLAB-Python bridge

WhoBrokeTheBuild opened the pull request:
This is a first pass at supporting mdsthin as a backend for connecting MATLAB to MDSplus The mechanism for enabling it is an extension of how you enable python
```matlab
% Use Java
mdsInfo()
% Use Python
mdsInfo(true)
% Use Python with mdsthin
mdsInfo(true, true)
```

Of course, it will throw an error if you attempt to `mdsconnect('local')` while using mdsthin

WhoBrokeTheBuild committed:
Feature: Allow using mdsthin for the MATLAB-Python bridge

This is a first pass at supporting mdsthin as a backend for connecting MATLAB to MDSplus
The mechanism for enabling it is an extension of how you enable python
```
% Use Java
mdsInfo()
% Use Python
mdsInfo(true)
% Use Python with mdsthin
mdsInfo(true, true)
```

Of course, it will throw an error if you attempt to `mdsconnect('local')` while using mdsthin

WhoBrokeTheBuild commented on matlab/private/pythonActivate.m:
```diff
@@ -5,7 +5,11 @@
+        if MDSINFO.usemdsthin
+            MDSINFO.ispy2 = false
+        else
+            MDSINFO.ispy2 = logical(py.MDSplus.version.ispy2);
```
We should consider removing this check altogether, the last matlab version that supported python 2 was R2022b

mwinkel-dev replied:
I concur.

I hope there are no customers still using MATLAB R2022b.   Disabling this check would be a way to find out (grin).   Plus, it would be easy to undo if the user base clamors for Python2 and ancient MATLAB releases.

mwinkel-dev commented:
I have reviewed the changes and they are all straightforward.

However, I am holding off on the approval until I've installed this commit on my test systems and used it for a few days.

mwinkel-dev commented on matlab/mdsInfo.m:
```diff
@@ -48,7 +52,7 @@
         disp('Unable to connect to MDSplus using java bridge, using python bridge instead')
         disp(strcat(' Java error: ', err{2}.message))
     else % ~ok(1)
-        disp('Unable to connect to MDSplus using python bridge, using python bridge instead')
```
This message should end with `using java bridge instead`.

During the investigation of #2925, it was noticed that if the specified bridge is not available, then the code tries the alternate bridge.   It is important to notify the user that if the Python bridge fails to launch, that it is then running the Java bridge.

mwinkel-dev commented on matlab/mdsInfo.m:
```diff
@@ -8,11 +8,15 @@
 end
 if nargin > 0
     MDSINFO.usePython = logical(varargin{1});
+    if nargin > 1
```
To guard against bogus input that attempts to use `mdsthin` with the Java bridge . . .

 ```
mdsInfo(false, true)
```

. . . the `if` statement should probably be as follows.

```
if (nargin > 1) && MDSINFO.usePython
```

mwinkel-dev commented:
Although it is correct that `mdsInfo(true)` does enable the Python bridge, the documentation instead recommends that customers use the wrapper function, `mdsUsePython()`.

Perhaps the `mdsUsePython()` function should also accept a second parameter to specify use of `mdsthin` with Python.    Suggested changes for the `mdsUsePython.m` file follow.

```
   if nargin == 0
       mdsInfo(true, false);
   elseif nargin == 1
       mdsInfo(varargin{1});
   else
       mdsInfo(varargin{1}, varargin{2});
   end
```

mwinkel-dev commented:
I have confirmed that the `mdsthin-bridge` works fine on a M2 MacBook with MATLAB R2024b.   All it requires is the `$MDSPLUS_DIR/matlab` directory.

Here are the configuration steps . . . 

1. Install `mdsthin` via `pip`.
2. The `$MDSPLUS_DIR/matlab` directory was copied to a `thin_bridge` directory. 
3. MATLAB's search path was then modified so it no longer referenced `$MDSPLUS_DIR/matlab`, but instead referenced the `thin_bridge` directory.
4. Logged out and logged back in to get a clean shell environment (i.e., no references to $MDSPLUS_DIR or related environment variables).
5. Started MATLAB.   (Didn't need graphical output, so used `matlab  -nodisplay` because it launches faster.)
6. At the MATLAB `>>` prompt typed the following statements

```
mdsInfo(true, true)  % specifies the "mdsthin" bridge
mdsconnect("<server>")
mdsopen("cmod", 1090909009)
mdsvalue("\ip")
mdsdisconnect
```

Next step is to install the `mdsthin-bridge` on Arsene's M2 MacBook.

mwinkel-dev commented:
Note that `mdsInfo()` does not mean "use Java"; it merely returns the current `MDSINFO` structure.   Yes, that structure is initialized to use the Java-bridge.   However, if the user has switched to the Python-bridge and then wants to switch back to the Java-bridge, `mdsInfo()` won't do the job.   They must instead explicitly use `mdsInfo(false)` or `mdsUsePython(false)`.

mwinkel-dev commented on matlab/mdsInfo.m:
```diff
@@ -48,7 +52,7 @@
+        disp('Unable to connect to MDSplus using python bridge')
         disp(strcat(' Python error: ', err{1}.message))
     end
 end
```
If control flow makes it beyond the final `end`, then a bridge has been successfully opened.   In which case, the function should return the current state of the `MDSINFO` structure.

```
info = MDSINFO;
```

WhoBrokeTheBuild committed:
Add Mark's patch

Update mdsdisconnect to not call `mdsconnect('local')` for mdsthin

mwinkel-dev reviewed (dismissed):
Confirmed this PR has all the changes made in the prototype installed on Arsene's MacBook.   (His computer is Apple Silicon and has been running the mdsthin-bridge for a month or so with no problems.)  This PR also has a useful enhancement regarding the `mdsdiscconect.m` script (so it doesn't throw an error when disconnecting from the mdsthin-bridge).   Approving this PR.

mwinkel-dev commented:
The testing done two hours ago used the wrong version of the `mdsdisconnect.m` file (and probably other files too).   After updating to the current version of the files in this PR, other errors were also found.   A full retest will be done later today and the results posted to this PR.

mwinkel-dev commented on matlab/mdsconnect.m:
```diff
@@ -21,7 +21,11 @@
                 else
                     MDSINFO.connection.mdsdisconnect();
                 end
```
There should be another 'end'  here to close out the 'if MDSINFO.isConnected' block.   Without this additional 'end', none of the three bridges will connect to a server.

mwinkel-dev commented on matlab/mdsdisconnect.m:
```diff
@@ -2,6 +2,14 @@
-    status = mdsconnect('local');
+    global MDSINFO
+    if MDSINFO.usemdsthin
+        MDSINFO.connection = []
```
Best to have a trailing ";" on the "... connection = []" statement.

mwinkel-dev commented on matlab/mdsdisconnect.m:
```diff
@@ -2,6 +2,14 @@
+        MDSINFO.connection = []
+        MDSINFO.connectedHost = '';
+        MDSINFO.isConnected = false;
+        status = 1
```
The "status = 1" should also have a trailing ";" to suppress output to the screen.   Having the ";" is also consistent with the "else" clause below.

mwinkel-dev commented:
Thorough testing with the correct *.m files revealed some problems (noted in previous posts).

There is also a design question.   If using the mdsthin-bridge and then issuing a `mdsUsePython()`, should it remain in mdsthin mode or switch to the standard Python mode?   Current behavior is shown below.

```
>> mdsUsePython(true, true)
>> mdsInfo()

ans = 

  struct with fields:

           isConnected: 0
            connection: []
         connectedHost: 'local'
             usePython: 1
            usemdsthin: 1
    isPythonConnection: 0
                 ispy2: 0

>> mdsUsePython(true) 
>> mdsInfo()

ans = 

  struct with fields:

           isConnected: 0
            connection: []
         connectedHost: 'local'
             usePython: 1
            usemdsthin: 1
    isPythonConnection: 0
                 ispy2: 0

>> 
```

Unrelated to this PR, a pre-existing problem was also found. If issue `mdsclose()` when a "local" tree is open via Python, it returns error code 265388200 = FD180A8 = TreeNOT_OPEN. However, when using Java, `mdsclose()` correctly returns the TreeSUCCESS status.

WhoBrokeTheBuild committed:
Add the changes from Mark's review

Add an extra `, false` in `mdsUsePython` to resolve the edge case Mark found when switching from the regular python bridge to the mdsthin one

mwinkel-dev reviewed (dismissed):
The third commit to this PR fixes the problems encountered with the previous commit.   The manual tests all pass, thus approving this PR.

The new documentation should explain how to install the mdsthin-bridge for MATLAB.   (Some users don't want a full MDSplus installation; they just want the simplest way to get MATLAB to read data from an MDSplus archive server.)  The easiest installation is probably to do the following:

1. install mdsthin,
2. download the appropriate *.tgz file from GitHub, and
3. extract just the `mdsplus/matlab` directory.

WhoBrokeTheBuild commented:
We could consider creating an extra tarball just for the matlab code, if that would be helpful to people

WhoBrokeTheBuild commented:
I added some code to the Jenkinsfile to create the matlab tgz, did a test build with it, and it looks pretty good to me. I'll attach it here (Just strip off the .txt, GitHub didn't like me uploading a tgz).
[mdsplus_PR-2935_7.156.0_matlab.tgz.txt](https://github.com/user-attachments/files/23126800/mdsplus_PR-2935_7.156.0_matlab.tgz.txt)

WhoBrokeTheBuild committed:
Add matlab package for GitHub

This is for use with the new mdsthin bridge, which doesn't require the rest of MDSplus to be installed

mwinkel-dev reviewed (approved):
A separate *.tgz file for the mdsthin-bridge is an excellent idea!   Thanks for changing the Jenkins file accordingly.   It all looks good, so approving this PR.
